torch/csrc/jit/script/compiler.cpp

```diff
@@ -29,6 +29,114 @@ using ValueTable = std::unordered_map<std::string, SugaredValuePtr>;
 
+
+using TypeAndRange = std::pair<TypePtr, const SourceRange*>;
+struct Refinements {
+
+  //using ordered map for deterministic graph output
+  std::map<std::string, TypeAndRange> mappings;
+
+  void setRefinement(const std::string& name, TypeAndRange mapping) {
+    mappings[name] = std::move(mapping);
+  }
+
+  c10::optional<TypeAndRange> getRefinement(const std::string& name) const {
+    const auto& maybe_mapping = mappings.find(name);
+    if (maybe_mapping == mappings.end()) {
+      return c10::nullopt;
+    }
+    return maybe_mapping->second;
+  }
+
+  // return the intersection of the values to type mappings in a and b whose
+  // types can be unified
+  Refinements* intersectRefinements(const Refinements& other) {
+    Refinements ret;
+    for (auto& name_mapping: mappings) {
+      const auto& name = name_mapping.first;
+      const auto& mapping = name_mapping.second;
+      if (auto other_mapping = other.getRefinement(name_mapping.first)) {
+        auto maybe_unified_type = unifyTypes(mapping.first, other_mapping->first);
+        if (maybe_unified_type) {
+          ret.setRefinement(name, TypeAndRange(*maybe_unified_type, mapping.second));
+        }
+      }
+    }
+    mappings = std::move(ret.mappings);
+    return this;
+  }
+
+  // return the union of the values to type mappings in a and b whose
+  // types can be unified
+  Refinements* unionRefinements(const Refinements& other) {
+    const Refinements& a = *this;
+    Refinements ret;
+    for (auto& name_mapping: mappings) {
+      const auto& name = name_mapping.first;
+      const auto& mapping = name_mapping.second;
+      TypePtr t_1 = mapping.first;
+      if (auto other_mapping = other.getRefinement(name_mapping.first)) {
+        TypePtr t_2 = other_mapping->first;
+        c10::optional<TypePtr> maybe_unified_type = c10::nullopt;
+        if (t_1->isSubtypeOf(t_2)) {
+          maybe_unified_type = t_1;
+        } else if (t_2->isSubtypeOf(t_1)) {
+          maybe_unified_type = t_2;
+        }
+        if (maybe_unified_type) {
+          ret.setRefinement(name, TypeAndRange(*maybe_unified_type, mapping.second));
+        }
+      } else {
+        ret.setRefinement(name, mapping);
+      }
+    }
+
+    for (auto& name_mapping: other.mappings) {
+      if (!a.getRefinement(name_mapping.first)) {
+        ret.setRefinement(name_mapping.first, name_mapping.second);
+      }
+    }
+
+    mappings = std::move(ret.mappings);
+    return this;
+  }
+};
+
+// When a comparison like x is None is made, we associate type refinements
+// with its true value and its false value. If a boolean that has refinements
+// associated with it is used in a conditional of an if statememt, the true and
+// false refinements are inserted into the corresponding blocks
+
+struct BoolInfo {
+  BoolInfo(Refinements true_refinements, Refinements false_refinements)
+      : true_refinements(std::move(true_refinements)),
```

Comment:
add `_` for both member variables. This constructor demonstrates why it's important: if you ever used `true_refinements` or `false_refinements` in the body of the constructor you'd be accessing a moved-from value.